Updates Artifacts Landing Page (#627)
wandb/docs: docs/guides/artifacts/intro.md

<CTAButtons productLink="https://wandb.ai/wandb/arttest/artifacts/model/iv3_trained/5334ab69740f9dda4fed/lineage" colabLink="https://colab.research.google.com/github/wandb/examples/blob/master/colabs/wandb-artifacts/Pipeline_Versioning_with_W%26B_Artifacts.ipynb"/>

Use W&B Artifacts to track and version any serialized data as the inputs and outputs of your [W&B Runs](../runs/intro.md). For example, a model training run might take in a dataset as input and trained model as output. In addition to logging hyper-parameters and metadata to a run, you can use an artifact to log the dataset used to train the model as input and the resulting model checkpoints as outputs.

> ## Use cases

Reviewer: can add a breakdown of a few common input/output combos on the use cases they represent? We often only talk about using a dataset as input and a model as output, and people get the idea that only a training run should have its artifacts tracked. Here are some other ideas to add in this use cases section:
1. ML workflow (run type) = dataset processing / transformation (input = dataset, output = dataset)
2. ML workflow (run type) = model evaluation or inference (input(s) = dataset, model, output = table)
3. ML workflow (run type) = model quantization (or putting model in different format) input = model, output = model). 

The goal is to convey the idea that you can have a W&B project that doesn't contain any training runs - for example we have some customers that just have a 'dataset factory' project where all inputs and output artifacts are datasets, and they're trying to track the lineage of transformations or processing run on their datasets like de-duplication, removal of PII, sampling, etc.)

Reviewer: Hey @katjacksonWB - sharing a diagram which shows some of these combinations created by phil a while back
![image (40)](https://github.com/wandb/docodile/assets/40642416/f9bd6123-71f4-4297-afbb-6eb39a240293)

Reviewer: You can see here some examples of the variety of workflows a user might want to add reproducibility for using a (1) run (2) input artifact (3) output artifact. My ask here is to highlight some other workflows aside from the most common one we promote - Model Training (which you can see above in the center) which takes in a dataset version and outputs a model version. 

For example - Data Prep (which aligns with my #1 above):
> ML workflow (run type) = dataset processing / transformation (input = dataset, output = dataset)

I would probably call this dataset pre-processing or transformation though instead of dataset prep. 

Another example above is model evaluation - this takes a model as input and outputs a table artifact to inspect results. 

My proposal would be to highlight 4 use cases for artifacts:
1. Model Training (most basic one - dataset artifact as input, model artifact as output)
2. Dataset Pre-Processing (dataset artifact as input, dataset artifact as output)
3. Model Evaluation (model artifact + test dataset artifact as input, table artifact as output)
4. Model Optimization (model artifact as input, model artifact as output)

Reviewer: Once you have another iteration on this specific section around use cases - I'm thinking we can loop in Ryan McConville (SMLE link for Artifacts) to get his eyes on whether the way these use cases are articulated make sense to him!

Reviewer: Hey @katjacksonWB is this resolved?

You can use artifacts throughout your entire ML workflow as inputs and outputs of [runs](../runs/intro.md):

![](/images/artifacts/artifacts_landing_page2.png)

This means that you can:

> * [View where a model came from, including trained data](./explore-and-traverse-an-artifact-graph.md).

Reviewer: ```suggestion
* [Visualize a model's or dataset's lineage and dependencies](./explore-and-traverse-an-artifact-graph.md).
```

<CTAButtons productLink="https://wandb.ai/wandb/arttest/artifacts/model/iv3_trained/5334ab69740f9dda4fed/lineage" colabLink="https://colab.research.google.com/github/wandb/examples/blob/master/colabs/wandb-artifacts/Pipeline_Versioning_with_W%26B_Artifacts.ipynb"/>

Use W&B Artifacts to track and version any serialized data as the inputs and outputs of your [W&B Runs](../runs/intro.md). For example, a model training run might take in a dataset as input and trained model as output. In addition to logging hyper-parameters and metadata to a run, you can use an artifact to log the dataset used to train the model as input and the resulting model checkpoints as outputs.

## Use cases
You can use artifacts throughout your entire ML workflow as inputs and outputs of [runs](../runs/intro.md). You can use datasets, models, or even other artifacts as inputs for processing.

![](/images/artifacts/artifacts_landing_page2.png)

> Some of the use cases of artifacts including:

Reviewer: Hey @katjacksonWB - as discussed can we redo this section on more focus on inputs + outputs? This is an example table for automations I had made (very different use case) which is one way/idea to present the combinations of inputs and outputs and run types. This automations had a similar goal of showing what all the options are for the type of workflows you can harness artifacts for. 
<img width="862" alt="Screenshot 2024-04-02 at 5 22 39 PM" src="https://github.com/wandb/docodile/assets/40642416/7e254464-9d3d-4dde-885f-e205e1444a96">

Author: Will do- that table's a super good reference point. Thank you!

<CTAButtons productLink="https://wandb.ai/wandb/arttest/artifacts/model/iv3_trained/5334ab69740f9dda4fed/lineage" colabLink="https://colab.research.google.com/github/wandb/examples/blob/master/colabs/wandb-artifacts/Pipeline_Versioning_with_W%26B_Artifacts.ipynb"/>

> Use W&B Artifacts to track and version data as the inputs and outputs of your [W&B Runs](../runs/intro.md). For example, a model training run might take in a dataset as input and trained model as output. In addition to logging hyperparameters and metadata to a run, you can use an artifact to log the dataset used to train the model as input and the resulting model checkpoints as outputs.

Reviewer 2: Suggesting a few changes below..

Use W&B Artifacts to track and version data as the inputs and outputs of your [W&B Runs](../runs/intro.md). For example, a model training run might take in a dataset as input and produce a trained model as output. In addition to logging hyperparameters, metadata and metrics to a run, you can use an artifact to log, track and version the dataset used to train the model as input and another artifact for the resulting model checkpoints as outputs.

## Use cases
You can use artifacts throughout your entire ML workflow as inputs and outputs of [runs](../runs/intro.md). You can use datasets, models, or even other artifacts as inputs for processing.

![](/images/artifacts/artifacts_landing_page2.png)

| Use Case               | Input                       | Output                       |
|------------------------|-----------------------------|------------------------------|

> | Model Training         | Dataset (training data)     | Trained [Model](../models.md)                |

Reviewer 2: If not too much, I would suggest modifying under "Input" to be _Dataset (training and validation data)_.

| Dataset Pre-Processing | Dataset (raw data)          | Dataset (pre-processed data) |

> | Model Evaluation       | Model + Dataset (test data) | [Table](../tables/intro.md)                        |

Reviewer 2: Perhaps specify this as a "W&B Table"?

| Model Optimization     | Model                       | Optimized Model              |

## Create an artifact

Create an artifact with four lines of code:

> 1. Create a [W&B run](../runs/intro.md).

Reviewer 2: Should this be W&B Run (i.e. uppercase)?

4. Log your artifact to W&B.

```python
run = wandb.init(project="artifacts-example", job_type="add-dataset")
artifact = wandb.Artifact(name="my_data", type="dataset")
artifact.add_dir(local_path="./dataset.h5")  # Add dataset directory to artifact
run.log_artifact(artifact)  # Logs the artifact version "my_data:v0"
```

:::tip

> See the [track external files](./track-external-files.md) page for information on how to add references to files or directories stored in external object storage, like an Amazon S3 bucket. 

Reviewer: ```suggestion
See the [track external files](./track-external-files.md) page for information on how to add references to files or directories stored in external object storage, like an Amazon S3 bucket. 
```
can we add a note explaining when this would be a relevant feature to explore? i.e - "have all your data stored in an external storage bucket? You can leave it all in there and just create an artifact that points to your stored content. We can still get checksum info and track version changes without having any access to your content / having you directly upload it to W&B."

Reviewer: Hey @katjacksonWB is this resolved?

:::

## Download an artifact
Indicate the artifact you want to use with the [`use_artifact`](../../ref/python/run.md#use_artifact) method, which returns a run object:

```python
artifact = run.use_artifact("my_data") #returns a run object using the "my_data" artifact
```

Then, use the returned object to download all contents of the artifact:

```python
datadir = artifact.download() #downloads the "my_data" artifact
```

:::tip
For alternate ways to download artifacts and to see additional parameters, see the guide on [downloading and using artifacts](./download-and-use-an-artifact.md)
:::

> ## Next steps

Reviewer: can we add a pointer to automation docs? i.e want to automate actions based off changes to your artifacts? 

Thinking similar for model registry - lets at a CTA to model registry docs.

Reviewer: Hey @katjacksonWB is this resolved?

4. Log your artifact to W&B.

```python
run = wandb.init(project="artifacts-example", job_type="add-dataset")
artifact = wandb.Artifact(name="my_data", type="dataset")
artifact.add_dir(local_path="./dataset.h5")  # Add dataset directory to artifact
run.log_artifact(artifact)  # Logs the artifact version "my_data:v0"
```

:::tip
See the [track external files](./track-external-files.md) page for information on how to add references to files or directories stored in external object storage, like an Amazon S3 bucket. 
:::

## Download an artifact

> Indicate the artifact you want to use with the [`use_artifact`](../../ref/python/run.md#use_artifact) method, which returns a run object:

Reviewer: ```suggestion
Indicate the artifact you want to mark as input to your run with the [`use_artifact`](../../ref/python/run.md#use_artifact) method, which returns an artifact object:
```
Possible mistake - I think this returns an artifact object not run

```python

> artifact = run.use_artifact("my_data") #returns a run object using the "my_data" artifact

Reviewer: is the input param format correct? doesn't it have to have version specified via :latest?

```

Then, use the returned object to download all contents of the artifact:

```python

> datadir = artifact.download() #downloads the "my_data" artifact

Reviewer: can we update the comment to tell the user specifically what they can expect to see when calling download? i.e creates a dir called artifacts/ locally with contents XYZ... so user know where on their machine this contents will show up

```

Optionally, you can retrieve the artifact's URL for reference:

```python

> print("reference file:", artifact.manifest.entries[ref].ref) #prints a URL pointing to the "my_data" artifact for future reference

Reviewer: I think this might be better to put inside the reference artifacts doc no? It's kind of confusing to have this when we're not talking about reference artifacts only in this section

```

:::tip
For alternate ways to download artifacts and to see additional parameters, see the guide on [downloading and using artifacts](./download-and-use-an-artifact.md)
:::

## Next steps
* Learn how to [version](./create-a-new-artifact-version.md), [update](./update-an-artifact.md), or [delete](./delete-artifacts.md) artifacts.

> * Learn about [artifact automation](./project-scoped-automations.md).

Reviewer: for the artifact automation can we be specific on why / how artifact automations connects? My suggestion was that we have something like: "To learn how to trigger downstream workflows in response to changes to your artifacts, go to artifact automations"

> 4. Log your artifact to W&B.

Reviewer 2: "Add one or more files or directories.."? It's useful to know files can be added individually, or a directory (and the example code is adding a directory).

```python
run = wandb.init(project="artifacts-example", job_type="add-dataset")
artifact = wandb.Artifact(name="my_data", type="dataset")

> artifact.add_dir(local_path="./dataset.h5")  # Add dataset directory to artifact

Reviewer 2: `artifact.add_dir(local_path="./dataset.h5")` appears a little confusing to me because dataset.h5 implies it's a file, not a directory, but the code calls `add_dir()`. Should it be `add_file()` or instead using `add_dir()` with something that is more clearly a directory?

Author: @rymc Thanks for the comments! I think I addressed all of them; would you mind taking another look? Thanks!

Reviewer 2: Looks good to me, thanks @katjacksonWB!
